clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -234,10 +234,15 @@ public RuntimeException lastError() {
     }
 
     public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
-        if (isInTransaction() && !partitionsInTransaction.contains(tp)) {
-            transitionToFatalError(new IllegalStateException("Attempted to dequeue a record batch to send " +
-                    "for partition " + tp + ", which hasn't been added to the transaction yet"));
-            return false;
+        if (isInTransaction() || hasError()) {
```

Comment:
That's a good point. In general, I guess there is no point retrying produce requests if you are fenced or have lost your permissions. And in a fatal error, we have to close and restart the producer, resetting all the sequence numbers. So we should always return false if the error is fatal. 

Will update.